src/explorer/connectionTreeItem.ts

```diff
@@ -95,21 +97,37 @@ export default class ConnectionTreeItem extends vscode.TreeItem
       return Promise.resolve([]);
     }
 
-    if (this._childrenCacheIsUpToDate) {
+    const dataService = this._connectionController.getActiveDataService();
+    if (dataService === null) {
+      return Promise.reject(new Error('Not currently connected.'));
+    }
+
+    if (this.cacheIsUpToDate) {
+      const pastChildrenCache = this._childrenCache;
+      this._childrenCache = {};
+
+      // We create a new database tree item here instead of reusing the
+      // cached one in order to ensure the expanded state is set.
+      Object.keys(pastChildrenCache).forEach(databaseName => {
+        this._childrenCache[databaseName] = new DatabaseTreeItem(
+          databaseName,
+          dataService,
+          pastChildrenCache[databaseName].isExpanded,
+          pastChildrenCache[databaseName].cacheIsUpToDate,
+          pastChildrenCache[databaseName].getChildrenCache()
+        );
+      });
+
       return Promise.resolve(Object.values(this._childrenCache));
     }
 
     return new Promise((resolve, reject) => {
-      const dataService = this._connectionController.getActiveDataService();
-      if (dataService === null) {
-        return reject(new Error('Not currently connected.'));
-      }
-      dataService.listDatabases((err: any, databases: string[]) => {
+      dataService.listDatabases((err: Error | undefined, databases: any[]) => {
         if (err) {
           return reject(new Error(`Unable to list databases: ${err.message}`));
         }
 
-        this._childrenCacheIsUpToDate = true;
+        this.cacheIsUpToDate = true;
```

Comment:
How is this used?

Reply:
We store the `cacheIsUpToDate` value on tree items with asynchronous resources in order to avoid re-fetching the resource when the tree changes. The cache is used to persist the expanded state of its children and their resources, this makes it so that when a connection is collapsed and expanded and list databases is re-run, we can keep the previous expanded state of databases and their collections.

Hmmm... We could use the cache as the flag here and set it to null when it needs to be refreshed. I think that might clean up the constructors of these classes a bit. Some of them have a few more arguments than I like.